src/main/java/com/google/api/codegen/config/GrpcGapicRetryMapping.java

```diff
@@ -0,0 +1,163 @@
+    defaultParamsBuilder.setTotalTimeoutMillis(0);
+    defaultParamsBuilder.setName("no_retry_params");
+    paramsDefMap.putIfAbsent(defaultParamsBuilder.getName(), defaultParamsBuilder.build());
+
+    RetryCodesDefinitionProto.Builder defaultCodesBuilder = RetryCodesDefinitionProto.newBuilder();
+    defaultCodesBuilder.setName("no_retry_codes");
+    codesDefMap.putIfAbsent(defaultCodesBuilder.getName(), defaultCodesBuilder.build());
+
+    // build retry-to-interface mapping from gRPC ServiceConfig
+    int retryNdx = 1;
+    int noRetryNdx = 1;
+    for (com.google.api.codegen.grpc.MethodConfig methodConfig : config.getMethodConfigList()) {
+      RetryPolicy retryPolicy = methodConfig.getRetryPolicy();
+      String policyName;
+      if (methodConfig.getRetryOrHedgingPolicyCase() == RetryOrHedgingPolicyCase.RETRY_POLICY) {
+        policyName = "retry_policy_" + retryNdx++;
+      } else {
+        // make "unique" no_retry configs because the MethodConfig timeout may differ
+        policyName = "no_retry_" + noRetryNdx++;
+      }
+
+      long timeout = Durations.toMillis(methodConfig.getTimeout());
+
+      // construct retry params from RetryPolicy
+      RetryParamsDefinitionProto.Builder paramsBuilder = RetryParamsDefinitionProto.newBuilder();
+      paramsBuilder.setMaxRetryDelayMillis(Durations.toMillis(retryPolicy.getMaxBackoff()));
+      paramsBuilder.setInitialRetryDelayMillis(Durations.toMillis(retryPolicy.getInitialBackoff()));
+      paramsBuilder.setRetryDelayMultiplier(floatToDouble(retryPolicy.getBackoffMultiplier()));
+      paramsBuilder.setTotalTimeoutMillis(timeout);
+      paramsBuilder.setName(policyName + "_params");
+      paramsDefMap.putIfAbsent(paramsBuilder.getName(), paramsBuilder.build());
+
+      // construct retry codes from RetryPolicy
+      RetryCodesDefinitionProto.Builder codesBuilder = RetryCodesDefinitionProto.newBuilder();
+      codesBuilder.setName(policyName + "_codes");
+      retryPolicy
+          .getRetryableStatusCodesList()
+          .forEach(code -> codesBuilder.addRetryCodes(code.name()));
+      codesDefMap.putIfAbsent(codesBuilder.getName(), codesBuilder.build());
+
+      // apply specific method config (overwrites) or apply service config to all methods (does
+      // not overwrite method-specific policies)
+      for (Name name : methodConfig.getNameList()) {
+        String serviceBase = name.getService();
+
+        if (!Strings.isNullOrEmpty(name.getMethod())) {
+          String fullName = fullyQualifiedName(serviceBase, name.getMethod());
+          methodCodesMap.put(fullName, codesBuilder.getName());
+          methodParamsMap.put(fullName, paramsBuilder.getName());
+          continue;
+        }
+
+        Interface interProto = protoInterfaces.get(serviceBase);
+        for (Method method : interProto.getMethods()) {
+          String fullName = method.getFullName();
+          methodCodesMap.putIfAbsent(fullName, codesBuilder.getName());
```

Comment:
Yes, it is important in this context because this logic applies a Service-wide RetryPolicy to all of its Methods. However, a RetryPolicy for specifically named Method takes precedence over a Service-wide one. Thus, if a Method's RetryPolicy already exists in the mapping, we assume it was added by a named Method RetryPolicy.

Its like `policyA applies to google.cloud.langauge.v1.LangaugeService` while `policyB applies to google.cloud.language.v1.LangaugeService.AnnotateText`.

There is a comment at the top of the parent `for` loop describing this:
```
// apply specific method config (overwrites) or apply service config to all methods (does
// not overwrite method-specific policies)
```

Reply:
Ok, that  makes perfect sense, and also becomes a natural target for refactoring this big method: split the service-wide logic and method-wide logic (i.e. method-wide porition goes into a private method and is called here instead). If it is still big enough, we can put the internals of the for loop itself in another private method (so ther ewill be this method, calling loop internals  private method, which in its turn calls the method-wide logic private method).

The above is just a suggestion (please disregard it if you have other refactoring ideas or a solid reason to keep this method as big as it  already is).

Reply:
I'm not sure I follow the suggestion.
> split the service-wide logic and method-wide logic (i.e. method-wide porition goes into a private method and is called here instead)

The Method-specific logic should use `put()` to overwrite/take precedence, while the Service-wide logic should use `putIfAbsent()` so as to **not** overwrite an existing Method-specific policy. So it's not clear to me how that helper would be reused. Am I missing something?

If it is about reducing the line size of this function, I could also move the RetryPolicy-to-Retry(Params | Codes)Definition building logic into private helpers.

Reply:
I have refactored it. I pulled the Retry(Params | Codes)Definition building logic into private helpers and moved the `for` loop that applies the retry policy names to a method/service name into a private helper to make it read a bit better. 

@vam-google WDYT?

Reply:
@noahdietz Looks good to me. If we did not need the `putIfAbsent` stuff, then it could look prettier (instead of passing each map separately, we could pass one big builder of the final object (which aggreates the three maps builder), thus keeping the list of parameters of the private methods shorter). But it is all ok.

My concern was not that much about the size (sometimes it is just ok to have big methods, actually I think the refactored version is even slightly bigger then original). It was more about ensuring that each method does one concrete thing.